File: embedded/src/alarm/intensity.go
package alarm

import "math"

// applyIntensityCurve calculates the current intensity based on curve type and progress
func applyIntensityCurve(curve IntensityCurve, progress float64) int {
	switch curve.Curve {
	case "linear":
		return calculateLinearIntensity(curve, progress)
	case "asymptotic":
		return calculateAsymptoticIntensity(curve, progress)
	case "s-curve":
		return calculateSCurveIntensity(curve, progress)
	case "custom":
		return calculateCustomIntensity(curve, progress)
	default:
		return curve.StartIntensity
	}
}

func calculateLinearIntensity(curve IntensityCurve, progress float64) int {
	return int(float64(curve.StartIntensity) + 
		(float64(curve.EndIntensity-curve.StartIntensity) * progress))
}

func calculateAsymptoticIntensity(curve IntensityCurve, progress float64) int {
	param := 4.0 // Default value from frontend
	if curve.HyperParameter != nil {
		param = float64(*curve.HyperParameter)
	}

	if param != 0 {
		numerator := 1.0 - math.Exp(-progress*param*0.4)
		denominator := 1.0 - math.Exp(-param*0.4)
		factor := numerator / denominator
		return int(float64(curve.StartIntensity) + 
			float64(curve.EndIntensity-curve.StartIntensity) * factor)
	}
	
	// Fall back to linear if param is 0
	return calculateLinearIntensity(curve, progress)
}

func calculateSCurveIntensity(curve IntensityCurve, progress float64) int {
	param := 10.0 // Default value from frontend
	if curve.HyperParameter != nil {
		param = float64(*curve.HyperParameter)
	}

	numerator := math.Exp(param*progress) - 1
	denominator := (math.Exp(0.5*param) - 1) * (1 + math.Exp(param*(progress-0.5)))
	factor := numerator / denominator

	return int(float64(curve.StartIntensity) + 
		float64(curve.EndIntensity-curve.StartIntensity) * factor)
}

func calculateCustomIntensity(curve IntensityCurve, progress float64) int {
	if len(curve.ControlPoints) == 0 {
		return calculateLinearIntensity(curve, progress)
	}

	// Create points array with start and end points
	points := make([]Point, 0, len(curve.ControlPoints)+2)
	points = append(points, Point{X: 0, Y: float64(curve.StartIntensity)})
	points = append(points, curve.ControlPoints...)
	points = append(points, Point{X: 100, Y: float64(curve.EndIntensity)})

	// Find the segment where progress falls
	progress100 := progress * 100
	for i := 0; i < len(points)-1; i++ {
		if progress100 >= points[i].X && progress100 <= points[i+1].X {
			// Get points for Catmull-Rom calculation
			p0 := points[max(0, i-1)]
			p1 := points[i]
			p2 := points[i+1]
			p3 := points[min(len(points)-1, i+2)]

			// Calculate t in [0,1] for the current segment
			t := (progress100 - p1.X) / (p2.X - p1.X)

			// Catmull-Rom calculation
			t2 := t * t
			t3 := t2 * t

			y := 0.5 * (
				(2*p1.Y) +
				(-p0.Y+p2.Y)*t +
				(2*p0.Y-5*p1.Y+4*p2.Y-p3.Y)*t2 +
				(-p0.Y+3*p1.Y-3*p2.Y+p3.Y)*t3)

			return int(y)
		}
	}

	return curve.EndIntensity
}

// Helper functions
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: embedded/src/alarm/intensity.py
import math

# Linear interpolation
def linear(start, end, t):
    return start + (end - start) * t

# S-curve (logistic)
def s_curve(start, end, t, sharpness=10):
    # t in [0,1], sharpness controls steepness
    s = 1 / (1 + math.exp(-sharpness * (t - 0.5)))
    s0 = 1 / (1 + math.exp(-sharpness * (-0.5)))
    s1 = 1 / (1 + math.exp(-sharpness * (0.5)))
    norm = (s - s0) / (s1 - s0)
    return start + (end - start) * norm

# Asymptotic (exponential approach)
def asymptotic(start, end, t, decay=5):
    # t in [0,1], decay controls how fast it approaches end
    return start + (end - start) * (1 - math.exp(-decay * t)) / (1 - math.exp(-decay))

# Catmull-Rom spline for custom curve
def catmull_rom(points, t):
    # points: list of (x, y), t in [0,1]
    n = len(points)
    if n < 2:
        return points[0][1] if n == 1 else 0
    # Find segment
    seg = min(int(t * (n - 1)), n - 2)
    t0 = seg / (n - 1)
    t1 = (seg + 1) / (n - 1)
    local_t = (t - t0) / (t1 - t0)
    # Get control points
    p0 = points[max(seg - 1, 0)][1]
    p1 = points[seg][1]
    p2 = points[seg + 1][1]
    p3 = points[min(seg + 2, n - 1)][1]
    # Catmull-Rom formula
    return 0.5 * (
        2 * p1 +
        (-p0 + p2) * local_t +
        (2*p0 - 5*p1 + 4*p2 - p3) * local_t**2 +
        (-p0 + 3*p1 - 3*p2 + p3) * local_t**3
    )

def get_intensity(curve_type, start, end, t, hyper=None, points=None):
    if curve_type == 0:  # Linear
        return linear(start, end, t)
    elif curve_type == 1:  # Asymptotic
        return asymptotic(start, end, t, decay=hyper or 5)
    elif curve_type == 2:  # S-curve
        return s_curve(start, end, t, sharpness=hyper or 10)
    elif curve_type == 3 and points:  # Custom
        return catmull_rom(points, t)
    else:
        return linear(start, end, t)
